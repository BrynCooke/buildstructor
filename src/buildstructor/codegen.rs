use crate::buildstructor::utils::{
    AngleBracketedGenericArgumentsExt, ExprTupleExt, GenericsExt, IdentExt, TypeExt, TypeTupleExt,
};
use crate::lower::{FieldType, Ir};
use proc_macro2::{Ident, TokenStream};
use quote::{format_ident, quote};
use syn::punctuated::Punctuated;
use syn::{Expr, ExprCall, GenericArgument, Generics, Index, Result, Type, WhereClause};
extern crate inflector;
use inflector::Inflector;

pub fn codegen(ir: Ir) -> Result<TokenStream> {
    let module_name = &ir.module_name;
    let target_name = &ir.target_name;

    let (impl_generics, ty_generics, where_clause) = &ir.generics.split_for_impl();

    let param_generics = ir.param_generics();

    let builder_generics =
        Generics::combine(vec![&ir.generics, &ir.method_generics, &param_generics]);
    let (builder_impl_generics, _, builder_where_clause) = builder_generics.split_for_impl();

    let builder_generics_tuple = Generics::combine(vec![&ir.generics, &ir.method_generics]);
    let builder_tuple_ty_generics = builder_generics_tuple
        .to_generic_args()
        .insert(0, Type::Tuple(param_generics.to_tuple_type()));

    let all_generics = Generics::combine(vec![
        &ir.builder_generics,
        &ir.generics,
        &ir.method_generics,
    ]);
    let (_, all_ty_generics, _) = all_generics.split_for_impl();

    let method_generics = &ir.method_generics;
    let builder_init_generics = Generics::combine(vec![&ir.generics, &ir.method_generics]);
    let target_generics_raw: Vec<GenericArgument> = builder_init_generics
        .to_generic_args()
        .args
        .into_iter()
        .collect();

    let constructor_method_name = &ir.constructor_method_name;
    let constructor_args = ir.constructor_args();
    let constructor_return = &ir.return_type;
    let builder_method_name = &ir.builder_method_name;
    let builder_name = &ir.builder_name;
    let builder_methods = builder_methods(&ir, builder_where_clause)?;
    let builder_state_type_initial = ir.builder_state_type_initial();
    let builder_state_initial = ir.builder_state_initial();

    let async_token = ir.is_async.then(|| quote! {async});
    let await_token = ir.is_async.then(|| quote! {.await});
    let vis = &ir.vis;

    Ok(quote! {
        impl #impl_generics #target_name #ty_generics #where_clause {
            #vis fn #builder_method_name #method_generics() -> #module_name::#builder_name<#builder_state_type_initial, #(#target_generics_raw), *> {
                #module_name::new()
            }
        }

        mod #module_name {
            use super::*;

            pub fn new #builder_init_generics() -> #builder_name<#builder_state_type_initial, #(#target_generics_raw), *>
            {
                #builder_name {
                    fields: (#(#builder_state_initial ,) *),
                    _phantom: core::default::Default::default()
                }
            }

           pub struct __Required<T> {
                _phantom: std::marker::PhantomData<T>,
            }
            pub struct __Optional<T> {
                lazy: Option<T>,
            }
            pub struct __Set<T> {
                value: T,
            }

            fn __set<T>(value: T) -> __Set<T> {
                __Set { value }
            }

            fn __required<T>() -> __Required<T> {
                __Required::<T> {
                    _phantom: core::default::Default::default(),
                }
            }

            fn __optional<T>() -> __Optional<T> {
                __Optional::<T> {
                    lazy: None,
                }
            }

            impl<T: Default> From<__Optional<T>> for __Set<T> {
                fn from(o: __Optional<T>) -> Self {
                    __Set {
                        value: o.lazy.unwrap_or_default(),
                    }
                }
            }


            pub struct #builder_name #all_ty_generics {
                fields: __P,
                _phantom: (#(core::marker::PhantomData<#target_generics_raw>,) *)
            }

            #(#builder_methods)*

            impl #builder_impl_generics #builder_name #builder_tuple_ty_generics #builder_where_clause {
                pub #async_token fn build(self) #constructor_return {
                    #target_name::#constructor_method_name(#(#constructor_args),*) #await_token
                }
            }
        }
    })
}

pub fn builder_methods(
    ir: &Ir,
    builder_where_clause: Option<&WhereClause>,
) -> Result<Vec<TokenStream>> {
    let builder_generics = Generics::combine(vec![&ir.generics, &ir.method_generics]);
    Ok(ir.builder_fields
        .iter()
        .enumerate()
        .map(|(idx, f)| {
            let builder_name = &ir.builder_name;
            let field_name = &f.name;
            let method_name = format_ident!("{}", f.name);
            let ty = &f.ty;
            let builder_type_generics = ir.builder_type_generics();
            let params_before = match f.field_type {
                FieldType::Regular =>
                    builder_type_generics
                        .to_tuple_type()
                        .with_type(idx, f.ty.clone().wrap_in_generic(format_ident!("__Required"))),
                _ => builder_type_generics
                    .to_tuple_type()
                    .with_type(idx, f.ty.clone().wrap_in_generic(format_ident!("__Optional"))),

            };
            let params_after = match f.field_type {
                FieldType::Regular | FieldType::Option =>
                    builder_type_generics
                        .to_tuple_type()
                        .with_type(idx, f.ty.clone().wrap_in_generic(format_ident!("__Set"))),
                _ => builder_type_generics
                    .to_tuple_type()
                    .with_type(idx, f.ty.clone().wrap_in_generic(format_ident!("__Optional"))),
            };
            let before = builder_generics
                .to_generic_args()
                .insert(0, Type::Tuple(params_before));
            let after = builder_generics
                .to_generic_args()
                .insert(0, Type::Tuple(params_after));

            let set = call(format_ident!("__set"), vec![Expr::Path(field_name.to_expr_path())]);
            let new_state = params(ir, idx, field_name, &builder_type_generics, set);
            let set_some = call(format_ident!("__set"), vec![call(format_ident!("Some"), vec![Expr::Path(field_name.to_expr_path())])]);

            let builder_type_generics = Generics::combine(vec![&builder_type_generics.without(idx), &builder_generics]);

            match f.field_type {
                FieldType::Option => {
                    let and_method_name = format_ident!("and_{}", f.name);
                    let new_state_option = params(ir, idx, field_name, &builder_type_generics, set_some);
                    let mut field_collection_type = f.collection_type.clone();
                    let mut into_generics = None;
                    let mut into_call = None;
                    if f.collection_into {
                        let into_type = field_collection_type.replace(Type::parse("__T"));
                        let _ = into_generics.insert(Some(quote! {
                            <__T: Into<#into_type>>
                        }));
                        into_call = Some(quote!{
                            .into()
                        })
                    }
                    quote! {
                        impl #builder_type_generics #builder_name #before {
                            pub fn #method_name #into_generics(self, #field_name: #field_collection_type) -> #builder_name #after #builder_where_clause {
                                #builder_name {
                                    fields: #new_state_option,
                                    _phantom: core::default::Default::default()
                                }
                            }
                            pub fn #and_method_name #into_generics(self, #field_name: #ty) -> #builder_name #after #builder_where_clause {
                                #builder_name {
                                    fields: #new_state,
                                    _phantom: core::default::Default::default()
                                }
                            }
                        }
                    }
                },
                FieldType::Set => {
                    let (singular, plural) = single_plural_names(field_name);
                    let mut field_collection_type = f.collection_type.clone();
                    let mut into_generics = None;
                    let mut into_call = None;
                    if f.collection_into {
                        let into_type = field_collection_type.replace(Type::parse("__T"));
                        let _ = into_generics.insert(Some(quote! {
                            <__T: Into<#into_type>>
                        }));
                        into_call = Some(quote!{
                            .into()
                        })
                    }
                    let index = Index::from(idx);
                    quote! {
                        impl #builder_type_generics #builder_name #before {

                            pub fn #plural (mut self, #field_name: #ty) -> #builder_name #before #builder_where_clause {
                                self.fields.#index.lazy.get_or_insert_with(||core::default::Default::default()).extend(#field_name.into_iter());
                                self
                            }

                            pub fn #singular #into_generics(mut self, value: #field_collection_type) -> #builder_name #before #builder_where_clause{
                                self.fields.#index.lazy.get_or_insert_with(||core::default::Default::default()).insert(value #into_call);
                                self
                            }

                        }
                    }
                },
                FieldType::Vec => {
                    let (singular, plural) = single_plural_names(field_name);
                    let mut field_collection_type = f.collection_type.clone();
                    let mut into_generics = None;
                    let mut into_call = None;
                    if f.collection_into {
                        let into_type = field_collection_type.replace(Type::parse("__T"));
                        let _ = into_generics.insert(Some(quote! {
                            <__T: Into<#into_type>>
                        }));
                        into_call = Some(quote!{
                            .into()
                        })
                    }
                    let index = Index::from(idx);

                    quote! {
                        impl #builder_type_generics #builder_name #before {

                            pub fn #plural (mut self, #field_name: #ty) -> #builder_name #before #builder_where_clause{
                                self.fields.#index.lazy.get_or_insert_with(||core::default::Default::default()).extend(#field_name.into_iter());
                                self
                            }

                            pub fn #singular #into_generics(mut self, value: #field_collection_type) -> #builder_name #before #builder_where_clause{
                                self.fields.#index.lazy.get_or_insert_with(||core::default::Default::default()).push(value #into_call);
                                self
                            }

                        }
                    }
                },
                FieldType::Map => {
                    let (singular, plural) = single_plural_names(field_name);
                    let mut field_key_type = f.key_type.clone();
                    let mut field_value_type = f.value_type.clone();
                    let mut into_generics = Vec::new();
                    let mut field_key_into_call = None;
                    let mut field_value_into_call = None;
                    if f.key_into {
                        let into_type = field_key_type.replace(Type::parse("__K"));
                        let _ = into_generics.push(quote! {
                            __K: Into<#into_type>
                        });
                        field_key_into_call = Some(quote!{
                            .into()
                        })
                    }
                    if f.value_into {
                        let into_type = field_value_type.replace(Type::parse("__V"));
                        let _ = into_generics.push(quote! {
                            __V: Into<#into_type>
                        });
                        field_value_into_call = Some(quote!{
                            .into()
                        })
                    }

                    let into_generics_final = if into_generics.is_empty() {
                        None
                    }
                    else {
                        Some(quote! {
                            <#(#into_generics),*>
                        })
                    };


                    let index = Index::from(idx);
                    quote! {
                        impl #builder_type_generics #builder_name #before {

                            pub fn #plural (mut self, #field_name: #ty) -> #builder_name #before #builder_where_clause{
                                self.fields.#index.lazy.get_or_insert_with(||core::default::Default::default()).extend(#field_name.into_iter());
                                self
                            }

                            pub fn #singular #into_generics_final (mut self, key: #field_key_type, value: #field_value_type) -> #builder_name #before {
                                self.fields.#index.lazy.get_or_insert_with(||core::default::Default::default()).insert(key #field_key_into_call, value #field_value_into_call);
                                self
                            }
                        }
                    }
                },
                _ => quote! {
                    impl #builder_type_generics #builder_name #before {
                        pub fn #method_name (self, #field_name: #ty) -> #builder_name #after {
                            #builder_name {
                                fields: #new_state,
                                _phantom: core::default::Default::default()
                            }
                        }
                    }
                },
            }

        })
        .collect())
}

fn single_plural_names(ident: &Ident) -> (Ident, Ident) {
    let plural = format_ident!("{}", ident);
    let mut singular = format_ident!("{}", ident.to_string().to_singular());
    if plural == singular {
        singular = format_ident!("{}_entry", ident);
    }
    (singular, plural)
}

fn call(name: Ident, params: Vec<Expr>) -> Expr {
    Expr::Call(ExprCall {
        attrs: Default::default(),
        func: Box::new(Expr::Path(name.to_expr_path())),
        paren_token: Default::default(),
        args: Punctuated::from_iter(params),
    })
}

fn params(
    ir: &Ir,
    idx: usize,
    field_name: &Ident,
    builder_type_generics: &Generics,
    set: Expr,
) -> Expr {
    Expr::Tuple(
        builder_type_generics
            .to_expr_tuple(|idxp, _| {
                if idx == idxp {
                    Expr::Path(field_name.to_expr_path())
                } else {
                    ir.tuple_field(idxp)
                }
            })
            .with_expr(idx, set),
    )
}

#[cfg(test)]
mod tests {
    use crate::analyze::analyze;
    use crate::buildstructor::tests::*;
    use crate::codegen::codegen;
    use crate::lower::lower;

    macro_rules! assert_codegen {
        ($input:expr) => {
            let models = analyze($input).expect("Analysis failed");
            for model in models {
                let ir = lower(model).expect("Ir failed");
                if let Ok(codegen) = codegen(ir) {
                    if let Ok(new_ast) = syn::parse2(codegen.clone()) {
                        let output = prettyplease::unparse(&new_ast);
                        insta::assert_snapshot!(output);
                    } else {
                        panic!("Failed to generate valid code:\n{}", codegen);
                    }
                } else {
                    panic!("Failed generate code");
                }
            }
        };
    }

    #[test]
    fn single_field_test() {
        assert_codegen!(single_field_test_case());
    }

    #[test]
    fn pub_test() {
        assert_codegen!(pub_test_case());
    }

    #[test]
    fn multi_field_test() {
        assert_codegen!(multi_field_test_case());
    }

    #[test]
    fn generic_test() {
        assert_codegen!(generic_test_case());
    }

    #[test]
    fn async_test() {
        assert_codegen!(async_test_case());
    }

    #[test]
    fn fallible_test() {
        assert_codegen!(fallible_test_case());
    }

    #[test]
    fn into_test() {
        assert_codegen!(into_test_case());
    }

    #[test]
    fn into_where_test() {
        assert_codegen!(into_where_test_case());
    }

    #[test]
    fn option_test() {
        assert_codegen!(option_test_case());
    }

    #[test]
    fn collection_test() {
        assert_codegen!(collections_test_case());
    }

    #[test]
    fn collection_generics_test() {
        assert_codegen!(collections_generics_test_case());
    }

    #[test]
    fn collection_option_test() {
        assert_codegen!(collections_option_test_case());
    }

    #[test]
    fn returns_self_test() {
        assert_codegen!(returns_self_test_case());
    }

    #[test]
    fn multiple_generics_test() {
        assert_codegen!(multiple_generics_test_case());
    }

    #[test]
    fn collection_generics_test2() {
        assert_codegen!(collections_generics_test_case2());
    }
}
