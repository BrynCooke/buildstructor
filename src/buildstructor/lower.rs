use crate::analyze::ConstrutorModel;
use crate::buildstructor::utils::{IdentExt, PunctuatedExt, TypeExt};
use proc_macro2::{Ident, TokenStream};
use quote::{format_ident, quote};
use std::default::Default;
use syn::punctuated::Punctuated;
use syn::{
    Expr, ExprField, FnArg, GenericArgument, GenericParam, Generics, Index, Member, Pat, Result,
    ReturnType, Type, TypeParam, TypeTuple, Visibility,
};
use try_match::try_match;

pub struct Ir {
    pub module_name: Ident,
    pub target_name: Ident,
    pub builder_name: Ident,
    pub builder_fields: Vec<BuilderField>,
    pub constructor_name: Ident,
    pub constructor_method_name: Ident,
    pub return_type: ReturnType,
    pub is_async: bool,
    pub vis: Visibility,
    pub generics: Generics,
    pub builder_generics: Generics,
    pub method_generics: Generics,
    pub builder_method_name: Ident,
}

pub struct BuilderField {
    pub name: Ident,
    pub ty: Type,
    pub field_type: FieldType,
    pub key_type: Option<Type>,
    pub value_type: Option<Type>,
    pub collection_type: Option<Type>,
}

#[derive(Debug)]
pub enum FieldType {
    Regular,
    Optional,
    Vec,
    Set,
    Map,
}

pub fn lower(model: ConstrutorModel) -> Result<Ir> {
    Ok(Ir {
        vis: model.vis.clone(),
        module_name: format_ident!(
            "__{}_{}_builder",
            model.ident.to_string().to_lowercase(),
            model.constructor_name.to_string().to_lowercase()
        ),
        target_name: model.ident.clone(),
        builder_name: format_ident!("__{}Builder", model.ident.clone()),
        constructor_method_name: model.constructor_name.clone(),
        builder_method_name: builder_method_name(&model),
        builder_fields: builder_fields(&model),
        constructor_name: format_ident!("{}Constructor", model.ident.to_string()),
        return_type: builder_return_type(model.output, model.ident),
        is_async: model.is_async,
        generics: model.generics,
        builder_generics: Ir::builder_generics(),
        method_generics: model.method_generics,
    })
}

fn builder_return_type(mut return_type: ReturnType, target: Ident) -> ReturnType {
    if let ReturnType::Type(_, ty) = &mut return_type {
        let self_type = Box::new(Type::Path(format_ident!("Self").to_type_path()));
        if ty == &self_type {
            *ty = Box::new(Type::Path(target.to_type_path()));
        }
    }
    return_type
}

fn builder_fields(model: &ConstrutorModel) -> Vec<BuilderField> {
    model
        .args
        .iter()
        .filter_map(|f| match f {
            FnArg::Typed(t) => {
                let ident = try_match!(&*t.pat, Pat::Ident(x)=>x).ok()?;
                let field_type = field_type(&*t.ty);
                let args = t.ty.generic_args();
                let (key_type, value_type, collection_type) = match (
                    &field_type,
                    args.and_then(|args| args.iter().next()),
                    args.and_then(|args| args.iter().nth(1)),
                ) {
                    (
                        FieldType::Vec | FieldType::Set,
                        Some(GenericArgument::Type(collection_type)),
                        None,
                    ) => (None, None, Some(collection_type.clone())),
                    (
                        FieldType::Map,
                        Some(GenericArgument::Type(key_type)),
                        Some(GenericArgument::Type(value_type)),
                    ) => (Some(key_type.clone()), Some(value_type.clone()), None),
                    _ => (None, None, None),
                };

                Some(BuilderField {
                    ty: *t.ty.clone(),
                    name: ident.ident.clone(),
                    field_type,
                    key_type,
                    value_type,
                    collection_type,
                })
            }
            FnArg::Receiver(_) => None,
        })
        .collect()
}

fn builder_method_name(model: &ConstrutorModel) -> Ident {
    format_ident!(
        "{}builder",
        model
            .constructor_name
            .to_string()
            .strip_suffix("new")
            .expect("already checked that the method ends with new, qed")
    )
}

fn field_type(ty: &Type) -> FieldType {
    match ty.raw_ident() {
        Some(f) if f == format_ident!("Option") => FieldType::Optional,
        Some(f) if f == format_ident!("Vec") => FieldType::Vec,
        Some(f) if f == format_ident!("HashSet") => FieldType::Set,
        Some(f) if f == format_ident!("BTreeSet") => FieldType::Set,
        Some(f) if f == format_ident!("HashMap") => FieldType::Map,
        Some(f) if f == format_ident!("BTreeMap") => FieldType::Map,
        _ => FieldType::Regular,
    }
}

impl Ir {
    pub fn constructor_args(&self) -> Vec<TokenStream> {
        self.builder_fields
            .iter()
            .enumerate()
            .map(|(idx, _)| {
                let idx = Index::from(idx);
                quote! {
                    self.fields.#idx.into().value
                }
            })
            .collect()
    }

    pub fn tuple_field(&self, idx: usize) -> Expr {
        Expr::Field(ExprField {
            attrs: vec![],
            base: Box::new(Self::fields()),
            dot_token: Default::default(),
            member: Member::Unnamed(Index::from(idx)),
        })
    }

    pub fn fields() -> Expr {
        Expr::Field(ExprField {
            attrs: vec![],
            base: Box::new(Expr::Path(format_ident!("self").to_expr_path())),
            dot_token: Default::default(),
            member: Member::Named(format_ident!("fields")),
        })
    }

    pub fn builder_state_type_initial(&self) -> Type {
        Type::Tuple(TypeTuple {
            paren_token: Default::default(),
            elems: Punctuated::from_iter(self.builder_fields.iter().map(|field| {
                match field.field_type {
                    FieldType::Regular => field.ty.wrap_in_generic_with_module(
                        &self.module_name,
                        format_ident!("__Required"),
                    ),
                    _ => field.ty.wrap_in_generic_with_module(
                        &self.module_name,
                        format_ident!("__Optional"),
                    ),
                }
            }))
            .with_trailing(),
        })
    }

    pub fn builder_state_initial(&self) -> Vec<TokenStream> {
        self.builder_fields
            .iter()
            .map(|field| match field.field_type {
                FieldType::Regular => quote! {__required()},
                _ => quote! {__optional()},
            })
            .collect()
    }

    pub fn param_generics(&self) -> Generics {
        Generics {
            params: Punctuated::from_iter(self.builder_fields.iter().enumerate().map(
                |(idx, f)| {
                    format_ident!("__P{}", idx).to_generic_param(Some(
                        &f.ty
                            .wrap_in_generic(format_ident!("__Set"))
                            .wrap_in_generic(format_ident!("Into")),
                    ))
                },
            )),
            ..Default::default()
        }
    }

    fn builder_generics() -> Generics {
        Generics {
            params: Punctuated::from_iter(vec![format_ident!("__P").to_generic_param(None)]),
            ..Default::default()
        }
    }

    pub fn builder_type_generics(&self) -> Generics {
        Generics {
            params: Punctuated::from_iter(
                self.builder_fields.iter().enumerate().map(|(idx, _f)| {
                    GenericParam::Type(TypeParam::from(format_ident!("__{}", idx)))
                }),
            ),
            ..Default::default()
        }
    }
}
