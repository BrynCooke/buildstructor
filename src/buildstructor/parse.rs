use proc_macro2::TokenStream;
use syn::parse::{Parse, ParseStream};
use syn::{parse2, ItemImpl, Result};

#[derive(Clone, Debug)]
pub struct Ast {
    pub item: ItemImpl,
}

impl Parse for Ast {
    fn parse(input: ParseStream) -> Result<Self> {
        Ok(Ast {
            item: input.parse()?,
        })
    }
}

pub fn parse(item: TokenStream) -> Result<Ast> {
    parse2::<Ast>(item)
}

#[cfg(test)]
mod tests {
    use crate::buildstructor::tests::{multi_field_test_case, single_field_test_case};

    #[test]
    fn single_field_test_case_is_valid() {
        single_field_test_case();
    }

    #[test]
    fn multiple_field_test_case_is_valid() {
        multi_field_test_case();
    }
}
