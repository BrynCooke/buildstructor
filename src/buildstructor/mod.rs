pub mod analyze;
pub mod codegen;
pub mod lower;
pub mod parse;
pub mod utils;

#[cfg(test)]
mod tests {
    use crate::parse::Ast;
    use syn::parse_quote;

    pub fn single_field_test_case() -> Ast {
        parse_quote!(
            #[builder]
            impl Foo {
                fn new(simple: usize) -> Foo {
                    Self { simple }
                }
            }
        )
    }

    pub fn pub_test_case() -> Ast {
        parse_quote!(
            #[builder]
            impl Foo {
                pub fn new(simple: usize) -> Foo {
                    Self { simple }
                }
            }
        )
    }

    pub fn multi_field_test_case() -> Ast {
        parse_quote!(
            #[builder]
            impl Foo {
                fn new(simple: usize, simple2: usize) -> Foo {
                    Self { simple, simple2 }
                }
            }
        )
    }

    pub fn fallible_test_case() -> Ast {
        parse_quote!(
            #[builder]
            impl Foo {
                fn new(simple: usize) -> Result<Foo, String> {
                    Ok(Self { simple })
                }
            }
        )
    }

    pub fn async_test_case() -> Ast {
        parse_quote!(
            #[builder]
            impl Foo {
                async fn new(simple: usize) -> Foo {
                    Foo { simple }
                }
            }
        )
    }

    pub fn generic_test_case() -> Ast {
        parse_quote!(
            #[builder]
            impl<T> Foo<T> {
                fn new(simple: T) -> Foo<T> {
                    Self { simple }
                }
            }
        )
    }

    pub fn into_test_case() -> Ast {
        parse_quote!(
            #[builder]
            impl Foo {
                fn new<T: Into<String>>(simple: T) -> Foo {
                    Foo {
                        simple: simple.into(),
                    }
                }
            }
        )
    }

    pub fn into_where_test_case() -> Ast {
        parse_quote!(
            #[builder]
            impl Foo {
                fn new<T>(simple: T) -> Foo
                where
                    T: Into<String>,
                {
                    Foo {
                        simple: simple.into(),
                    }
                }
            }
        )
    }

    pub fn option_test_case() -> Ast {
        parse_quote!(
            #[builder]
            impl Foo {
                fn new(option: Option<usize>) -> Foo {
                    Foo { option }
                }
            }
        )
    }

    pub fn collections_test_case() -> Ast {
        parse_quote!(
            #[builder]
            impl Foo {
                fn new(
                    simple: usize,
                    set: HashSet<String>,
                    map: HashMap<String, String>,
                    vec: Vec<String>,
                    btmap: BTreeMap<String, String>,
                    btset: BTreeSet<String>,
                ) -> Foo {
                    Self {
                        simple,
                        set,
                        map,
                        vec,
                        btmap,
                        btset,
                    }
                }
            }
        )
    }
}
