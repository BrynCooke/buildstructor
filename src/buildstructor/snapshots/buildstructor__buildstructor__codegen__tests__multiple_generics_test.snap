---
source: src/buildstructor/codegen.rs
expression: output
---
impl<T> Request<T> {
    pub fn fake_builder<K, V>() -> __request_fake_new_builder::__RequestBuilder<
            (
                __request_fake_new_builder::__Optional<Vec<(K, V)>>,
                __request_fake_new_builder::__Optional<Option<http::Uri>>,
                __request_fake_new_builder::__Optional<Option<http::Method>>,
                __request_fake_new_builder::__Required<T>,
            ),
            T,
            K,
            V,
        > {
        __request_fake_new_builder::new()
    }
}
mod __request_fake_new_builder {
    use super::*;
    pub fn new<T, K, V>() -> __RequestBuilder<
            (
                __request_fake_new_builder::__Optional<Vec<(K, V)>>,
                __request_fake_new_builder::__Optional<Option<http::Uri>>,
                __request_fake_new_builder::__Optional<Option<http::Method>>,
                __request_fake_new_builder::__Required<T>,
            ),
            T,
            K,
            V,
        > {
        __RequestBuilder {
            fields: (__optional(), __optional(), __optional(), __required()),
            _phantom: core::default::Default::default(),
        }
    }
    pub struct __Required<T> {
        _phantom: std::marker::PhantomData<T>,
    }
    pub struct __Optional<T> {
        lazy: Option<T>,
    }
    pub struct __Set<T> {
        value: T,
    }
    fn __set<T>(value: T) -> __Set<T> {
        __Set { value }
    }
    fn __required<T>() -> __Required<T> {
        __Required::<T> {
            _phantom: core::default::Default::default(),
        }
    }
    fn __optional<T>() -> __Optional<T> {
        __Optional::<T> { lazy: None }
    }
    impl<T: Default> From<__Optional<T>> for __Set<T> {
        fn from(o: __Optional<T>) -> Self {
            __Set {
                value: o.lazy.unwrap_or_default(),
            }
        }
    }
    pub struct __RequestBuilder<__P, T, K, V> {
        fields: __P,
        _phantom: (
            core::marker::PhantomData<T>,
            core::marker::PhantomData<K>,
            core::marker::PhantomData<V>,
        ),
    }
    impl<
        __1,
        __2,
        __3,
        T,
        K,
        V,
    > __RequestBuilder<(__Optional<Vec<(K, V)>>, __1, __2, __3), T, K, V> {
        pub fn headers(
            mut self,
            headers: Vec<(K, V)>,
        ) -> __RequestBuilder<(__Optional<Vec<(K, V)>>, __1, __2, __3), T, K, V>
        where
            HeaderName: TryFrom<K>,
            <HeaderName as TryFrom<K>>::Error: Into<http::Error>,
            HeaderValue: TryFrom<V>,
            <HeaderValue as TryFrom<V>>::Error: Into<http::Error>,
        {
            self.fields
                .0
                .lazy
                .get_or_insert_with(|| core::default::Default::default())
                .extend(headers.into_iter());
            self
        }
        pub fn header(
            mut self,
            value: (K, V),
        ) -> __RequestBuilder<(__Optional<Vec<(K, V)>>, __1, __2, __3), T, K, V>
        where
            HeaderName: TryFrom<K>,
            <HeaderName as TryFrom<K>>::Error: Into<http::Error>,
            HeaderValue: TryFrom<V>,
            <HeaderValue as TryFrom<V>>::Error: Into<http::Error>,
        {
            self.fields
                .0
                .lazy
                .get_or_insert_with(|| core::default::Default::default())
                .push(value);
            self
        }
    }
    impl<
        __0,
        __2,
        __3,
        T,
        K,
        V,
    > __RequestBuilder<(__0, __Optional<Option<http::Uri>>, __2, __3), T, K, V> {
        pub fn uri(
            self,
            uri: http::Uri,
        ) -> __RequestBuilder<(__0, __Set<Option<http::Uri>>, __2, __3), T, K, V>
        where
            HeaderName: TryFrom<K>,
            <HeaderName as TryFrom<K>>::Error: Into<http::Error>,
            HeaderValue: TryFrom<V>,
            <HeaderValue as TryFrom<V>>::Error: Into<http::Error>,
        {
            __RequestBuilder {
                fields: (self.fields.0, __set(Some(uri)), self.fields.2, self.fields.3),
                _phantom: core::default::Default::default(),
            }
        }
        pub fn and_uri(
            self,
            uri: Option<http::Uri>,
        ) -> __RequestBuilder<(__0, __Set<Option<http::Uri>>, __2, __3), T, K, V>
        where
            HeaderName: TryFrom<K>,
            <HeaderName as TryFrom<K>>::Error: Into<http::Error>,
            HeaderValue: TryFrom<V>,
            <HeaderValue as TryFrom<V>>::Error: Into<http::Error>,
        {
            __RequestBuilder {
                fields: (self.fields.0, __set(uri), self.fields.2, self.fields.3),
                _phantom: core::default::Default::default(),
            }
        }
    }
    impl<
        __0,
        __1,
        __3,
        T,
        K,
        V,
    > __RequestBuilder<(__0, __1, __Optional<Option<http::Method>>, __3), T, K, V> {
        pub fn method(
            self,
            method: http::Method,
        ) -> __RequestBuilder<(__0, __1, __Set<Option<http::Method>>, __3), T, K, V>
        where
            HeaderName: TryFrom<K>,
            <HeaderName as TryFrom<K>>::Error: Into<http::Error>,
            HeaderValue: TryFrom<V>,
            <HeaderValue as TryFrom<V>>::Error: Into<http::Error>,
        {
            __RequestBuilder {
                fields: (
                    self.fields.0,
                    self.fields.1,
                    __set(Some(method)),
                    self.fields.3,
                ),
                _phantom: core::default::Default::default(),
            }
        }
        pub fn and_method(
            self,
            method: Option<http::Method>,
        ) -> __RequestBuilder<(__0, __1, __Set<Option<http::Method>>, __3), T, K, V>
        where
            HeaderName: TryFrom<K>,
            <HeaderName as TryFrom<K>>::Error: Into<http::Error>,
            HeaderValue: TryFrom<V>,
            <HeaderValue as TryFrom<V>>::Error: Into<http::Error>,
        {
            __RequestBuilder {
                fields: (self.fields.0, self.fields.1, __set(method), self.fields.3),
                _phantom: core::default::Default::default(),
            }
        }
    }
    impl<
        __0,
        __1,
        __2,
        T,
        K,
        V,
    > __RequestBuilder<(__0, __1, __2, __Required<T>), T, K, V> {
        pub fn body(
            self,
            body: T,
        ) -> __RequestBuilder<(__0, __1, __2, __Set<T>), T, K, V> {
            __RequestBuilder {
                fields: (self.fields.0, self.fields.1, self.fields.2, __set(body)),
                _phantom: core::default::Default::default(),
            }
        }
    }
    impl<
        T,
        K,
        V,
        __P0: Into<__Set<Vec<(K, V)>>>,
        __P1: Into<__Set<Option<http::Uri>>>,
        __P2: Into<__Set<Option<http::Method>>>,
        __P3: Into<__Set<T>>,
    > __RequestBuilder<(__P0, __P1, __P2, __P3), T, K, V>
    where
        HeaderName: TryFrom<K>,
        <HeaderName as TryFrom<K>>::Error: Into<http::Error>,
        HeaderValue: TryFrom<V>,
        <HeaderValue as TryFrom<V>>::Error: Into<http::Error>,
    {
        pub fn build(self) -> http::Result<Request<T>> {
            Request::fake_new(
                self.fields.0.into().value,
                self.fields.1.into().value,
                self.fields.2.into().value,
                self.fields.3.into().value,
            )
        }
    }
}

